spec/requests/users_spec.rb

```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+RSpec.describe "Users", type: :request do
+  describe "GET #new" do
+    it "has http status success" do
+      get "/welcome"
+      expect(response).to have_http_status(:success)
+    end
+  end
+
+  describe "POST #create" do
+    let(:params) {
+      {
+        user: { email: "foo@example.com", name: "Mickey Mouse" }
+      }
+    }
+
+    context "with new email" do
+      it "creates new User record" do
+        expect {
+          post "/welcome", params: params
+        }.to change(User, :count).by(1)
+
+        user = User.last
+        expect(user.email).to eq "foo@example.com"
+        expect(user.name).to eq "Mickey Mouse"
+      end
+
+      it "re-renders welcome page" do
+        post "/welcome", params: params
+        expect(response).to have_http_status(:success)
+        expect(response.body).to match "enter your email address"
+      end
+    end
+
+    context "with email already in use" do
+      let!(:user) { create(:user, email: "foo@example.com", name: "Foo Bar") }
+
+      it "does not create new user record" do
```

Comment:
One side note - I love the testing. We haven't talked about _what_ we're going to test so for now just go with what you're doing. It's instructive. 

This is a simple example where it doesn't really matter, but I see this request spec and think it does the same as the model spec for email validation. The page rendering is unique, but the `does not create new user record` should be covered if we trust the other spec, right? Or does this cover something that doesn't?

As this gets more complicated, we might target certain types of tests and not others is the only point. What do you think? Let's plan to discuss live in our next update, and we'll go from there.

Reply:
That's true -- the actual check for whether the user count increases we could probably skip, but I think the check that the name doesn't get updated is important.  I could have coded it that entering the same email twice but with a different name would have overwritten the original name.

Sure, we can discuss tests next time.  I generally try to have pretty good test coverage.